server/services/v1/auth.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/fullstorydev/grpchan/inprocgrpc"
+	"github.com/go-chi/chi/v5"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/server/config"
+	"google.golang.org/grpc"
+)
+
+const (
+	oauthPattern = "/oauth/*"
+	refreshToken = "refresh_token"
+	scope        = "offline_access openid"
+)
+
+type authService struct {
+	api.UnimplementedAuthServer
+	OAuthProvider
+}
+
+type OAuthProvider interface {
+	GetAccessToken(request *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error)
+}
+
+type Auth0Oauth struct {
+	config.AuthConfig
+}
+
+func (a Auth0Oauth) GetAccessToken(req *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error) {
+	data := url.Values{
+		"refresh_token": {req.RefreshToken},
+		"client_id":     {a.AuthConfig.ClientId},
+		"grant_type":    {refreshToken},
+		"scope":         {scope},
+	}
+	resp, err := http.PostForm(a.AuthConfig.ExternalTokenURL, data)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+	}
+	bodyStr := string(body)
+	if resp.StatusCode == 200 {
+		getAccessTokenResponse := api.GetAccessTokenResponse{}
+
+		err = json.Unmarshal([]byte(bodyStr), &getAccessTokenResponse)
+		if err != nil {
+			return nil, api.Errorf(api.Code_INTERNAL, "Failed to parse external response: reason = "+err.Error())
+		}
+		return &getAccessTokenResponse, nil
+	} else {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+bodyStr)
+	}
+}
+
+func newAuthService(authConfig config.AuthConfig) *authService {
+	if authConfig.OAuthProviderType == "auth0" {
+		return &authService{
+			OAuthProvider: Auth0Oauth{
+				authConfig,
+			},
+		}
+	}
+	log.Error().Msg("Unable to configure external oauth provider")
```

Comment:
add `.Str("type", authConfig.OAuthProviderType).` to the log.

Reply:
Actually, I am thinking if we should crash here?

Reply:
yes I think we should crash here too. updated log and updated it to crash for such exceptional configuration.